packages/test/test-end-to-end-tests/src/test/gcTests/gcUnreferencedFlagValidation.spec.ts

```diff
@@ -0,0 +1,306 @@
+            `Did not get expected summary. Expected: ${summarySequenceNumber}. ` +
+            `Actual: ${latestSummaryContext?.referenceSequenceNumber}.`,
+        );
+        assert(latestUploadedSummary !== undefined, "Did not get a summary");
+
+        // Validate the summary uploaded to the server
+        {
+            const dataStoreTrees =
+                (latestUploadedSummary.tree[channelsTreeName] as ISummaryTree)?.tree ?? latestUploadedSummary.tree;
+            for (const [key, value] of Object.entries(dataStoreTrees)) {
+                // The data store's summary will be a handle if it did not change since last summary. If so, ignore it.
+                if (value.type === SummaryType.Tree) {
+                    if (unreferencedDataStoreIds.includes(key)) {
+                        assert(value.unreferenced, `Data store ${key} should be marked as unreferenced in summary`);
+                    } else {
+                        assert(
+                            value.unreferenced === undefined,
+                            `Data store ${key} should not be marked as unreferenced in summary`,
+                        );
+                    }
+                }
+            }
+        }
+
+        // Validate the snapshot downloaded from the server.
+        {
+            // Download the snapshot corresponding to the above summary from the server.
+            const versions = await documentStorage.getVersions(summaryVersion, 1);
+            const snapshot = await documentStorage.getSnapshotTree(versions[0]);
+            // eslint-disable-next-line no-null/no-null
+            assert(snapshot !== null, "Snapshot could not be downloaded from server");
+            const dataStoreTrees = snapshot.trees[channelsTreeName]?.trees ?? snapshot.trees;
+            for (const [key, value] of Object.entries(dataStoreTrees)) {
+                if (unreferencedDataStoreIds.includes(key)) {
+                    assert(value.unreferenced, `Data store ${key} should be marked as unreferenced in snapshot`);
+                } else {
+                    assert(
+                        value.unreferenced === undefined,
+                        `Data store ${key} should not be marked as unreferenced in snapshot`,
+                    );
+                }
+            }
+        }
+    }
+
+    before(function() {
+        provider = getTestObjectProvider();
+        // Currently, only ODSP returns back the "unreferenced" flag in the snapshot. Once we add this to other
+        // servers, we should enable these tests for them too.
+        if (provider.driver.type !== "odsp") {
+            this.skip();
+        }
+    });
+
+    beforeEach(async () => {
+        // Wrap the document service factory in the driver so that the `uploadSummaryCb` function is called every
+        // time the summarizer client uploads a summary.
+        (provider as any)._documentServiceFactory = wrapDocumentServiceFactory(
```

Comment:
This is quite hacky. Couple of ways that I can think of to make this better:
- Make `documentServiceFactory` in TestObjectProvider writeable.
- Or I could skip validating the uploaded summary altogether and only validate the snapshot downloaded from server. This would make the tests much simpler but remove a layer of verification.

Reply:
I took a look at compatUtils.ts in test-version-utils to see what options we have here.

- I think making `documentServiceFactory` in TestObjectProvider writeable is dangerous, because the whole point of the compat utils is that it gives you a specific version of the driver, loader, etc. You could instead expose an optional wrapper/adapter to `describeNoCompat`, etc. functions so that it still uses the underlying driver code from its version. I imagine this would make the intention more clear- do not _replace_ functionality, but rather gives you a hook to do additional work/fetch additional data for testing purposes. I imagine this would look like an optional config parameter that accepts a `DocumentStorageServiceProxy` or similar (to leverage the constructor) that gets wrapped somewhere in the test setup code.
- For end-to-end tests, this makes more sense to me, when you say it removes a layer of verification, what does that mean? The only possible user data summary that the server could have would have to have come from this client, right? Additionally, as we expose more summary testing, we can probably find another way to achieve the first "layer" of verification.

Reply:
In the meantime, I think the hack is fine, since it's just a test.

I think you should make sure that the test will fail if we change the implementation of `provider` such that `_documentServiceFactory` no longer exists. But I think the way your tests are written, they should already fail, though it's indirect.

Reply:
I thought about adding param to configs but it would be very specific for only a small set of tests. So, making that generic might not justify the complexity.
What I mean by removes a layer of verification is that if the test fails then it could be one of the 2 reasons -
1. The client did not upload the `unreferenced` flag correctly in the summary.
2. The server did not persist / send the `ureferenced` flag correctly.
Verifying that the uploaded summary is correct / incorrect will make it easy to tell where the problem is.

Reply:
I am not sure what you mean in your second comment. When will `_documentServiceFactory` no longer exist? If that happens, everything will fail right?

Reply:
And yeah, the hack should be fine for now since with your `summarizer` testability changes, we won't need this hack anymore. So this is just temporary anyway.

Reply:
Sounds good to me. For the `_documentServiceFactory` part, I just meant that setting a new property would still work at runtime, but the tests would likely fail in "unexpected" ways, versus having an explicit check causing them to fail.